staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -121,7 +130,12 @@ func newOIDCAuthProvider(clusterAddress string, cfg map[string]string, persister
 
 	// Check cache for existing provider.
 	if provider, ok := cache.getClient(clusterAddress, issuer, clientID); ok {
-		return provider, nil
+		// Return the provider only if the refresh token hasn't changed. Otherwise,
+		// remove the stale client from the cache and let it re-initialize further
+		if cfg[cfgRefreshToken] == provider.cfg[cfgRefreshToken] {
+			return provider, nil
+		}
+		cache.deleteClient(clusterAddress, issuer, clientID)
```

Comment:
This is not thread safe, i.e. it suffers from TOCTOU.  More than one go routine could go into the `if` statement, and the second could unnecessarily delete the client.

Reply:
Thanks for pointing this out @enj! 

Probably the easiest way to fix it is to use `cache.mu.Lock` around both the `cache.getClient` and `cache.deleteClient` calls. 

But maybe it's better to use an approach suggested in https://github.com/kubernetes/kubernetes/pull/78509 ? Reuse the client from the cache but update its configuration? I think it's a slightly better solution than creating a long critical section here. WDYT?

Reply:
A side note from looking into this a bit more - the TOCTOU bug always there. Just imagine two Go Routines going parallel through the `if provider, ok := cache.getClient(clusterAddress, issuer, clientID)` check.

Reply:
I introduced a new mutex to and created a critical section of the `newOIDCAuthProvider` content.